SpatialGDK/Source/SpatialGDK/Private/Interop/InitialOnlyFilter.cpp

```diff
@@ -0,0 +1,136 @@
+	, Receiver(InReceiver)
+{
+}
+
+bool InitialOnlyFilter::HasInitialOnlyData(Worker_EntityId EntityId) const
+{
+	return (RetrievedInitialOnlyData.Find(EntityId) != nullptr);
+}
+
+bool InitialOnlyFilter::HasInitialOnlyDataOrRequestIfAbsent(Worker_EntityId EntityId)
+{
+	if (HasInitialOnlyData(EntityId))
+	{
+		return true;
+	}
+
+	if (InflightInitialOnlyEntities.Find(EntityId) != nullptr)
+	{
+		return false;
+	}
+
+	PendingInitialOnlyEntities.Add(EntityId);
+	return false;
+}
+
+void InitialOnlyFilter::FlushRequests()
+{
+	if (PendingInitialOnlyEntities.Num() == 0)
+	{
+		return;
+	}
+
+	TArray<Worker_Constraint> EntityConstraintArray;
+
+	for (auto EntityId : PendingInitialOnlyEntities)
+	{
+		UE_LOG(LogInitialOnlyFilter, Verbose, TEXT("Requested initial only data for entity %lld."), EntityId);
+
+		Worker_Constraint Constraints{};
+		Constraints.constraint_type = WORKER_CONSTRAINT_TYPE_ENTITY_ID;
+		Constraints.constraint.entity_id_constraint.entity_id = EntityId;
+
+		EntityConstraintArray.Add(Constraints);
+
+		InflightInitialOnlyEntities.Add(EntityId);
+	}
+
+	Worker_EntityQuery InitialOnlyQuery{};
+
+	InitialOnlyQuery.constraint.constraint_type = WORKER_CONSTRAINT_TYPE_OR;
+	InitialOnlyQuery.constraint.constraint.or_constraint.constraint_count = EntityConstraintArray.Num();
+	InitialOnlyQuery.constraint.constraint.or_constraint.constraints = EntityConstraintArray.GetData();
+	InitialOnlyQuery.snapshot_result_type_component_set_id_count = 1;
+	InitialOnlyQuery.snapshot_result_type_component_set_ids = &SpatialConstants::INITIAL_ONLY_COMPONENT_SET_ID;
+
+	const Worker_RequestId RequestID = Connection->SendEntityQueryRequest(&InitialOnlyQuery, SpatialGDK::RETRY_UNTIL_COMPLETE);
+	EntityQueryDelegate InitialOnlyQueryDelegate;
+	InitialOnlyQueryDelegate.BindRaw(this, &InitialOnlyFilter::HandleInitialOnlyResponse);
+
+	Receiver->AddEntityQueryDelegate(RequestID, InitialOnlyQueryDelegate);
+
+	InflightInitialOnlyRequests.Add(RequestID, { MoveTemp(PendingInitialOnlyEntities) });
+}
+
+void InitialOnlyFilter::HandleInitialOnlyResponse(const Worker_EntityQueryResponseOp& Op)
+{
+	ClearRequest(Op.request_id);
+
+	if (Op.status_code != WORKER_STATUS_CODE_SUCCESS)
+	{
+		UE_LOG(LogInitialOnlyFilter, Error, TEXT("Failed to retrieve initial only data. Code: %d, %s"), Op.status_code,
+			   UTF8_TO_TCHAR(Op.message));
+		return;
+	}
+
+	for (uint32_t i = 0; i < Op.result_count; ++i)
+	{
+		const Worker_Entity* Entity = &Op.results[i];
+		const Worker_EntityId EntityId = Entity->entity_id;
+
+		if (Connection->GetView().Find(EntityId) == nullptr)
+		{
+			UE_LOG(LogInitialOnlyFilter, Verbose, TEXT("Received initial only data for entity no longer in view. Entity: %lld."), EntityId);
+			continue;
+		}
+
+		UE_LOG(LogInitialOnlyFilter, Verbose, TEXT("Received initial only data for entity. Entity: %lld."), EntityId);
+
+		// Extract and store the initial only data.
+		TArray<ComponentData>& ComponentDatas = RetrievedInitialOnlyData.FindOrAdd(EntityId);
```

Comment:
Entity component data could already exist (hence the find or add), so little clumsy to attempt to reserve the additional space. Will leave as is unless you feel strongly.

Reply:
I suppose it would be, yes, you'll need to "Reserve(ComponentDatas.Num() + Entity->component_count)", and component_count could maybe be zero?

Only a nit, I don't feel too strongly about it.

Reply:
Yeah fair enough - made the change